README.md
# Pipeline Instrumentation and Visualization Tool (PIVT)

![UT MTTR](doc/images/ut_ttr.png) ![FT Results](doc/images/ft_results.png)
![VIC Status](doc/images/vic_status.png) ![DRs](doc/images/drs.png)

## Overview

PIVT is used to monitor the software factory. It gives much needed instrumentation and visualization of software development pipelines and tools. The dashboards provide useful information at multiple levels, including middle-level management, software team managers, and software engineers.

The backend is written in Python and Perl and pulls data from tools like Jenkins, ClearQuest, and BitBucket. This data is then fed into Splunk, where a PIVT app lives to serve as the frontend.

## Getting Started

### Hardware Requirements

The hardware specifications of the lab machine PIVT runs on in the Aerospace environment is as follows:
- 16 cores
- 8 GB RAM
- 500 GB storage

These are not hard requirements. We have not tested PIVT on a machine with less resources, so we don't have hard minimum requirements.
An educated guess would be
- 4 cores
- 4 GB RAM
- 500 GB storage

More users require more resources.

### Software Requirements

- Port 8089 on the machine running Splunk must be open if the scripts reside on a different machine
- Python version at least 3.5.2
- Python "requests" module installed
- Python "pyyaml" module installed
- Splunk version at least 7.0.2

### Installation

1. Create a Splunk user and role for the PIVT scripts to use.
	- Role:
		- name: script_user
		- inheritance: can_delete
		- capabilities: indexes_edit, list_inputs
	- User:
		- username: script_user
		- password: changeme

2. If data was included with the new delivery, delete and recreate all pivt_* indexes in Splunk. If not, skip this step.

3. From the PIVT home directory, run

```
./configure.sh <install-target>
./install.sh
```

configure.sh checks for the following:

- A valid Python version (3.5.2+)
- Python requests module installed
- An existing Splunk installation
- A valid Splunk version (7.0.2+)
- outputs a "config.out" file as input for install.sh

install.sh organizes PIVT files into the correct directory structure and installs the PIVT app.\
During installation, you will be asked to authenticate with Splunk.

#### Directory structure

PIVT is organized into several directories:

- `bin/` (scripts)
- `etc/` (configuration files)
- `install/` (installation files)
- `var/` (data and logs)

#### Notable files

##### bin/pivt/export_jenkins.py

Pulls data from Jenkins using the Jenkins REST API. Both instances, "Production" and "Development", are utilized.\
Specific URLs are loaded by the script from the file "etc/sources.txt". Each URL points to a specific CI + "stage" (i.e. "Build," "Deploy," etc.). In the future, these will reside in a configuration file.\
Data is in JSON format and stored in a timestamped directory in var/data/newdata.

##### bin/pivt/export_vic_status.py

Pulls data from the VIC status page using a REST API.
Data is in JSON format and stored in a timestamp directory in var/data/newdata.

##### bin/export_cq.pl

Exports data from ClearQuest using query in CQ database. This data is output to var/data/newdata.

##### bin/pivt/collect.py

Gathers all the data in var/data/newdata and zips it into one archive, placed in var/data/collected.

##### bin/pivt/process.py

Processes the raw data in var/data/collected.
Processed data goes to var/data/data.
Raw data in var/data/collected is moved to var/data/archive.
Requires a PIVT_HOME environment variable set.

##### install/pivt-splunk-app.tar.gz

The PIVT Splunk app contains the dashboards that allow a user to filter and view Jenkins and ClearQuest data graphically in Splunk.
The app will be installed by install.sh.

##### etc/meta.ini

Records the last pull time for each Jenkins job pulled by bin/pivt/export_jenkins.py.

### Notes

bin/ contains some shell scripts that make it easy to execute the Python/Perl scripts.
These should be used instead of calling the Python scripts directly and they should be called on a schedule.

They should be called in a particular sequence:
1. Export scripts
2. Collect script
3. Process script

Minimum desired call rate for export scripts:

- export_jenkins.py: every 6 hours
- export_vic_status.py: every hour
- export_cq.pl: every 24 hours

## License

[Apache 2.0](LICENSE)


CHANGELOG.md
# Changelog
All notable changes to PIVT will be documented in this file

## 0.1.2.19.1 - 2019-09-23

- Added FT (BETA) dashboard
- Changed name of RCR dashboard to Release Closure Review (BETA)

## 0.1.2.19 - 2019-09-12

- Changed CQ discovery rate graph to stacked area line chart
- Added DMS_BUILD_VERSION to RCR table
- Fixed RCR dashboard, ensuring EIS runs have an EIS_BUILD_VERSION
- Added build_passed column to RCR dashboard table
- Added new CQ fields to aggregate by list on CQ dashboard
- Added discovery rate pie chart to CQ dashboard
- Added pipeline run inspector dashboard
- Moved pipeline run tagging feature from RCR dashboard to pipeline run inspector
- Added drilldown capability to RCR dashboard for digging into pipeline runs

## 0.1.2.18 - 2019-08-05

- Added "FT Run Inspector" dashboard
- Added drilldown to FT Run Inspector dashboard from Jenkins Interactive and Go Green dashboards
- Removed "files changed" table from CQ dashboard
- Updated scenario master lists

## 0.1.2.17.1 - 2019-07-16

- Fixed time range issue on "Matching Core Scenarios" table on FT Truth dashboard
- Fixed mismatched numbers on "Truth Scenario Count by CI" and "Matching Core Scenarios" tables on FT Truth dashboard
- Added description text to "Matching Core Scenarios" table on FT Truth dashboard
- Set y-axis minimum to 0 on "Actual core tests run vs truth core count by CI" panel on FT Truth dashboard
- Renamed "All truth scenarios" panel to "Truth Table" and moved it to the bottom on FT Truth dashboard
- Fixed issue where %%ci33%% runs would not show if 0 stages executed on %%ci33%% and %%ci33%% Run Inspector dashboards
- Added skipped column to UT table on Jenkins Interactive
- Added "avg failed planned regression scenarios" column to FT tables on Jenkins Interactive and Go Green dashboards

## 0.1.2.17 - 2019-07-01

- Added "pipeline" and "branch" filters to %%ci33%% dashboard
- Added beta RCR dashboard
- Added "element" and "lab" filters to ClearQuest dashboard and replace "CI"
  filter with "Verifying Org"
- Demoted CQ dashboard to beta status
- Improved performance of dashboard that report on FTs
- Fixed auto install bug
- Fixed bug in CQ data processing that didn't allow for new fields in the
  schema

## 0.1.2.16 - 2019-06-17

- Started tracking remaining P-Release pipelines
- Started pulling all %%ci33%% pipeline jobs
- Promoted VIC Utilization dashboard
- Changed default time range from "last 30 days" to "last 7 days" for Go Green, Jenkins Production Statistics,
  and Pipeline Availability dashboards
- Enhanced performance of Internal Metrics dashboard
- Added panel on usage % by user on Internal Metrics dashboard
- Added %%ci33%% build type descriptions to Jenkins Interactive and Go Green dashboards
- Fixed bug in duration matrix on Jenkins Production Statistics dashboard
- Updated %%ci33%% dashboard with info on where pipeline started
- Added "timeline" vizualization to help visualize %%ci33%% building
- Updated FT truth table
- Updated reference run implementation to using Jenkins parameters
    - AWS_VIC=blank
    - CLEARCASE_VIEW=blank
    - Build=true
    - Deploy=true
    - Functional_Test=true
- Stopped taking skipped UTs into account for total on Jenkins Interactive dashboard
- Added %%ci33%% Run Inspector dashboard allowing drilldown on one %%ci33%% pipeline run
- Added drilldown capability to %%ci33%% dashboard table
- Fixed FT pass rate calculation issue on Jenkins Interactive

## 0.1.2.15 - 2019-05-13

- Added "runs by platform type" chart to Go Green dashboard
- Added matching scenarios table to FT Truth dashboard
- Added new "Runs" table to %%ci33%% dashboard
- Changed name of nominal filter to "Reference Run" on Jenkins Interactive dashboard
- Added reference run filter to Go Green dashboard
- Added pipeline filter to Jenkins Interactive and Go Green dashboards
- Added drilldown to Go Green dashboard
- Added %%ci21%% pipeline to sources
- Added column descriptions on Go Green

## 0.1.2.14 - 2019-04-15

- Started pulling pipeline.json file from Jenkins for future use
- Now using DEFAULT_BUILD Jenkins data field to determine if a run is "nominal"
- Added "nominal build" filter to Jenkins - Interactive and Go Green dashboards
- Updated VIC status URL
- Updated Python path in Raytheon scripts
- Fixed average scenario pass rate metric on Go Green dashboard statistics table
- Added summary row to VIC Utilization statistics table
- Added platform version filters to Jenkins - Interactive and Go Green dashboards
- Fixed bug where P-Release Jenkins jobs weren't getting pulled
- Renamed and moved some configuration files
- Added configuration management module; default configs (do not touch) are in etc/default whereas
  local configs (modifiable) are in etc/local
- pipeline.properties and pipeline.json files are now pulled for every Jenkins build
- Started counting all "skipped" functional tests as "failed"
- Removed legacy %%ci33%%_BUILD_TYPE determination that used PIPELINE_URL to determine if a P-Release or Blob image was used
- Started pulling PostCore and AllCores %%ci33%% jobs
- Fixed bug in Go Green "Successful FT Runs by CI over Time" chart
- Updated FT truth table
- Added new beta dashboard: "Internal Metrics (BETA)"

## 0.1.2.13 - 2019-02-27

- Changed PIVT delivery structure to use "dist" directory
- Many enhancements to auto install
- Added CI filter to VIC Utilization dashboard
- Added utilization table to VIC Utilization dashboard
- pivt.log now rotates each week
- Split VERSION file into two: VERSION and LASTPULL
- Added MANIFEST file to delivery

## 0.1.2.12 - 2019-02-08

- Enhance cucumber report processing to make it deterministic when there are duplicate scenarios
- Added min and max scenarios run columns to Go Green stats table
- Changed Go Green stats table to disregard runs with no pass/fail numbers
- Fixed bug that would cause a crash if the CQ index couldn't be deleted after processing data

pivt/util.py
# -*- coding: utf-8 -*-

# Copyright 2019 The Aerospace Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Utility functions
"""

import logging
import logging.handlers
import os
import sys
import time
import datetime
import re
import shutil
from urllib.request import urlopen
from collections import abc
from pathlib import Path
import ssl
import pivt.conf_manager as cm


class Utility:
    """Utility class to hold common values and functions."""
    def __init__(self):
        # Paths
        self.pivt_home = Path()

        self.etc_dir = Path()
        self.var_dir = Path()

        # config files
        self.version_file = Path()
        self.sources_file = Path()
        self.sources_file_ins = Path()
        self.sources_file_vic = Path()
        self.metadata_file = Path()
        self.unpulled_builds_file = Path()

        self.log_dir = Path()

        # data directories
        self.data_dir = Path()
        self.archive_dir = Path()
        self.collected_dir = Path()
        self.db_dir = Path()
        self.new_data_dir = Path()

        self.jenkins_data_dir = Path()
        self.jenkins_ft_data_dir = Path()
        self.ins_data_dir = Path()
        self.vic_data_dir = Path()
        self.cq_data_dir = Path()
        self.cq_data_path_old = Path()
        self.cq_changed_files_path = Path()
        self.cq_data_path = Path()
        self.cq_events_path = Path()
        self.vic_status_data_dir = Path()

        # Logging
        self.file_handler = None

        self.logger_formatter = logging.Formatter('%(asctime)s [%(name)s] %(levelname)s - %(message)s')

        console_handler = logging.StreamHandler()
        console_handler.setLevel(logging.INFO)
        console_handler.setFormatter(self.logger_formatter)

        root_logger = logging.getLogger('')
        root_logger.setLevel(logging.DEBUG)
        root_logger.addHandler(console_handler)

        self.logger = self.get_logger(self)

        self.splunk_url = ''

        # any Jenkins file pulled after this date without an instance field
        # in its data will have its instance set to "Production". else, it is "Development"
        self.first_default_prod_date = datetime.datetime(2018, 1, 23)

        self.ci_to_ss = Utility.setup_ci_to_ss()

        self.conf_manager = None

        self.initialized = False

    def setup(self):
        """
        Initialize
        """
        if self.initialized:
            return

        self.setup_env()

        self.file_handler = self.get_logger_file_handler('pivt')
        logging.getLogger('').addHandler(self.file_handler)

        self.splunk_url = 'https://{0}:8089'.format(os.environ.get('SPLUNK_HOST', 'localhost'))

        self.conf_manager = cm.ConfManager(self.etc_dir)
        self.conf_manager.load()

        self.initialized = True

    def setup_env(self):
        """Initialize environment."""
        if 'PIVT_HOME' not in os.environ:
            sys.exit('No PIVT_HOME in environment. Exiting.')

        self.pivt_home = Path(os.environ['PIVT_HOME'])

        if not self.pivt_home.exists():
            sys.exit('PIVT_PATH {0} does not exist. Exiting.'.format(self.pivt_home))

        self.etc_dir = self.pivt_home / 'etc'
        self.var_dir = self.pivt_home / 'var'

        self.version_file = self.etc_dir / 'pivt.version'  # file containing current and previous version info
        self.sources_file = self.etc_dir / 'product.sources'
        self.sources_file_ins = self.etc_dir / 'ins.sources'
        self.sources_file_vic = self.etc_dir / 'vic.sources'
        self.metadata_file = self.etc_dir / 'job_pull_times.ini'
        self.unpulled_builds_file = self.etc_dir / 'unpulled.json'

        self.log_dir = self.var_dir / 'log'

        self.data_dir = self.var_dir / 'data'  # path to directory with files monitored by Splunk
        self.archive_dir = self.data_dir / 'archive'  # path to archive directory
        self.collected_dir = self.data_dir / 'collected'  # path to collected directory
        self.db_dir = self.data_dir / 'data'
        self.new_data_dir = self.data_dir / 'newdata'

        self.jenkins_data_dir = self.db_dir / 'jenkins'
        self.jenkins_ft_data_dir = self.jenkins_data_dir / 'ft'

        self.ins_data_dir = self.db_dir / 'ins'

        self.vic_data_dir = self.db_dir / 'vic'

        self.cq_data_dir = self.db_dir / 'cq'
        self.cq_data_path_old = self.cq_data_dir / 'CQ_Data.csv'
        self.cq_changed_files_path = self.cq_data_dir / 'CQ_Changed_Files.csv'

        self.cq_data_path = self.cq_data_dir / 'drs.csv'
        self.cq_events_path = self.cq_data_dir / 'events.json'

        self.vic_status_data_dir = self.db_dir / 'vic_status'

        self.log_dir.mkdir(parents=True, exist_ok=True)

    def teardown(self):
        """Tear down the util instance. Mainly used for unit tests."""
        if self.file_handler is not None:
            self.file_handler.close()

        util.rmtree(self.log_dir, no_exist_ok=True)

        self.__init__()

    @staticmethod
    def setup_ci_to_ss():
        """
        Set up CI to subsystem map.
        """
        ss_to_ci = {
            '%%ci31%%': ['%%ci16%%', '%%ci17%%'],
            '%%ci32%%': ['%%ci18%%', '%%ci19%%', '%%ci15%%', '%%ci20%%', '%%ci12%%'],
            '%%ci21%%': ['%%ci21%%'],
            '%%ci22%%': ['%%ci22%%', '%%ci23%%', '%%ci24%%', '%%ci25%%', '%%ci9%%', '%%ci7%%', '%%ci8%%', '%%ci10%%'],
            '%%ci33%%': ['%%ci26%%'],
            '%%ci34%%': ['%%ci27%%'],
            '%%ci35%%': ['%%ci28%%', '%%ci11%%', '%%ci29%%'],
            '%%ci36%%': ['%%ci30%%', '%%ci13%%', '%%ci14%%']
        }

        ci_to_ss = {}

        for ss, cis in ss_to_ci.items():
            for ci in cis:
                ci_to_ss[ci] = ss

        return ci_to_ss

    @staticmethod
    def get_logger(obj, file_handler=None):
        """
        Create a logger.
        :param obj: the name of the logger or class that is requesting it
        :param file_handler: an optional file handler to add to the logger
        :return: the logger
        """
        if isinstance(obj, str):
            name = obj
        else:
            name = str(obj.__class__.__name__)

        logger = logging.getLogger(name)

        if file_handler is not None:
            logger.addHandler(file_handler)

        return logger

    def get_logger_file_handler(self, title, when='W6', backup_count=5):
        """
        Create a logger file handler.
        :param title: the name of the file
        :param when: frequency to rotate the log file
        :param backup_count: number of log files to keep
        :return: the file handler
        """
        file_handler = logging.handlers.TimedRotatingFileHandler(str((self.log_dir / (title + '.log'))), when=when, backupCount=backup_count)
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(self.logger_formatter)

        return file_handler

    @staticmethod
    def gen_jenkins_event_key(event, time_field='timestamp'):
        """
        Generate Jenkins event key.
        :param event: Jenkins event
        :param time_field: event timestamp field
        :return: Jenkins event key
        """
        return '{0}:{1}:{2}:{3}'.format(event['ci'], event['stage'], event['number'], event[time_field])

    @staticmethod
    def gen_ins_event_key(event, time_field='timestamp'):
        """
        Generate %%ci33%% event key.
        :param event: %%ci33%% event
        :param time_field: event timestamp field
        :return: %%ci33%% event key
        """
        return '{0}:{1}:{2}'.format(event['core'], event['id'], event[time_field])

    @staticmethod
    def gen_jenkins_project_key(project, number):
        """
        Returns a unique key for an event instance

        :param project: The project the event instance belongs to
        :param number:  The build number of the project
        :returns: A key that uniquely identifies an event instance
        """
        return '{0}:{1}'.format(project, number)

    @staticmethod
    def get_project_name(event):
        """
        Get a Jenkins event project name.
        :param event: the Jenkins event
        :return: the project name
        """
        try:
            full_display_name = event['fullDisplayName']
            return full_display_name.split(' ')[0]
        except KeyError:
            pass

        try:
            url = event['url']
            return url.split('/')[-3]
        except KeyError:
            return None

    def update_dashboards(self, splunk_home, logger, get_new_date=False):
        """
        Update Splunk dashboards with current version and pull date.
        :param splunk_home: path to Splunk installation
        :param logger: logger to use for logging
        :param get_new_date: if True, generate a new pull date using current time; else, load from file
        """
        new_description = self._get_new_dashboard_description(get_new_date)
        path = self._get_dashboards_path(splunk_home, logger)
        dashboard_names = os.listdir(path)
        for dashboard_name in dashboard_names:
            self._update_last_pull_date(dashboard_name, path, new_description, logger)

    def _get_dashboards_path(self, path_to_splunk, logger):
        """
        Get path to PIVT Splunk app dashboards.
        :param path_to_splunk:
        :param logger: the logger to use for logging
        :return: path to dashboards
        """
        if not logger:
            logger = self.logger

        dashboards_path = path_to_splunk + '/etc/apps/pivt/default/data/ui/views/'
        if not os.path.exists(dashboards_path):
            logger.info('Could not find pivt/default; using pivt/local')
            dashboards_path = path_to_splunk + '/etc/apps/pivt/local/data/ui/views/'
            if not os.path.exists(dashboards_path):
                raise FileNotFoundError('Could not find pivt/local')

        return dashboards_path

    def _update_last_pull_date(self, dashboard_name, dashboards_path, new_dashboard_description, logger):
        """
        Update description for a Splunk dashboard.
        :param dashboard_name: name of dashboard to update
        :param dashboards_path: path to dashboards in Splunk installation
        :param new_dashboard_description: description to update dashboard with
        :param logger: logger to use for logging
        """
        if not logger:
            logger = self.logger

        logger.info('Updating last pull date on %s dashboard', dashboard_name)

        path = dashboards_path + '/' + dashboard_name

        try:
            with open(path, 'r') as file:
                dashboard = file.read()

            dashboard = re.sub(r'<description>.*</description>\n', new_dashboard_description, dashboard)

            with open(path, 'w') as file:
                file.write(dashboard)
        except FileNotFoundError:
            logger.error('File does not exist: %s', path)

    def _get_new_dashboard_description(self, get_new_date=False):
        """
        Get a new dashboard description.
        :param get_new_date: if True, generate a new pull date using current time; else, load from file
        :return new description
        """
        version_string = self.read_value_from_kv_file(self.version_file, 'CURRENT')

        if not get_new_date:
            last_pull_date = self.conf_manager.get('pivt', 'general', 'last_pull')
        else:
            now_date_time = datetime.datetime.now(datetime.timezone.utc).astimezone()
            last_pull_date = self.get_normalized_now_date_time(now_date_time)
            self.conf_manager.set('pivt', 'general', 'last_pull', last_pull_date, create=True)

        return '<description>PIVT Version: {0} --- Last pull date: {1}</description>\n'.format(version_string, last_pull_date)

    @staticmethod
    def get_normalized_now_date_time(now_date_time):
        """
        Produce a normalized date/time string for Splunk dashboards.
        :param now_date_time: the current time in a datetime object
        :return: formatted date/time string
        """
        return '{0} {1}'.format(now_date_time.strftime('%x @ %X'), now_date_time.tzname())

    @staticmethod
    def read_value_from_kv_file(file_path, key):
        """
        Read the value of a key in a file. The file should have the following form:
            key1=value1
            key2=value2
            ...
        :param file_path: path to file to read from
        :param key: key to look up
        :return the value associated with key
        """
        if file_path.exists():
            with file_path.open() as file:
                for line in file:
                    line = line.strip()
                    pull_date_match = re.match(key + '=(.*)', line)
                    if pull_date_match:
                        return pull_date_match.group(1)

        return None

    @staticmethod
    def update_value_in_kv_file(file_path, key, value):
        """
        Update the value of the key in the file to be the value given as an argument.
        :param file_path: path to file to write to
        :param key: key of new value
        :param value: new value associated with key
        """
        new_kv = '{0}={1}\n'.format(key, value)

        if file_path.exists():
            with file_path.open() as file:
                file_data = file.read()

            new_file_data = re.sub(r'{0}=.*\n'.format(key), new_kv, file_data)
        else:
            new_file_data = new_kv

        with file_path.open('w') as file:
            file.write(new_file_data)

    @staticmethod
    def rmtree(path, no_exist_ok=False):
        """
        Safely remove a directory.
        :param path: path to directory to remove
        :param no_exist_ok: if True, don't error if the directory doesn't exist
        """
        if isinstance(path, Path):
            path = str(path)

        if str(path) == '.':
            return

        for i in range(0, 5):
            try:
                shutil.rmtree(path)
                break
            except FileNotFoundError:
                if not no_exist_ok:
                    raise
            except (OSError, PermissionError):
                if i >= 9:
                    raise
                time.sleep(0.2)

    @staticmethod
    def listdir(path):
        """
        Wrapper around os.listdir enabling it on Path objects
        :param path: path to list
        :return: result of listdir
        """
        return os.listdir(str(path))

    @staticmethod
    def basename(path):
        """
        Wrapper around os.path.basename enabling it on Path objects
        :param path: path to get basename from
        :return: basename
        """
        if isinstance(path, str):
            return os.path.basename(path)
        if isinstance(path, Path):
            return path.name

        raise Exception('path must be of type str or Path! path: ' + path)

    @staticmethod
    def inner_stringify(value):
        """
        Stringify inner elements of a dict or string
        :param value: object to stringify
        :return: stringified object
        """
        if isinstance(value, abc.Mapping):
            new_dictionary = {}
            for k, v in value.items():
                new_dictionary[str(k)] = Utility.inner_stringify(v)
            return new_dictionary

        if isinstance(value, abc.Sequence) and not isinstance(value, str):
            new_list = []
            for v in value:
                new_list.append(Utility.inner_stringify(v))
            return new_list

        return str(value)

    @staticmethod
    def get(url):
        """Read data from a URL"""
        return urlopen(url, context=ssl._create_unverified_context()).read()


class Constants:
    """Common constants"""
    SOLVED = 'solved'
    UNSOLVED = 'unsolved'
    CAUSE_NOT_ASSIGNED = 'Not Assigned'
    VERSION_NOT_ASSIGNED = 'Not Assigned'
    ITERATION_NOT_ASSIGNED = 'Not Assigned'
    ROOT_CAUSES = ['user', 'nightly', 'self-service', 'weekly']
    UNKNOWN_VERSION = 'N/A'
    UNIT_TEST = 'UnitTest'
    FUNCTIONAL_TEST = 'FunctionalTest'
    PRODUCTION = 'Production'
    DEVELOPMENT = 'Development'


util = Utility()
